lib/ex_doc/language/elixir.ex

```diff
@@ -0,0 +1,240 @@
+defmodule ExDoc.Language.Elixir do
+  @moduledoc false
+
+  @behaviour ExDoc.Language
+
+  alias ExDoc.Autolink
+
+  @impl true
+  def module_data(module) do
+    {type, skip} = module_type_and_skip(module)
+
+    %{
+      id: inspect(module),
+      title: module_title(module, type),
+      type: type,
+      skip: skip,
+      extra_callback_types: [:macrocallback]
+    }
+  end
+
+  @impl true
+  def function_data(entry, module_data) do
+    {{kind, name, arity}, _anno, _signature, _doc_content, metadata} = entry
+
+    extra_annotations =
+      case {kind, name, arity} do
+        {:macro, _, _} -> ["macro"]
+        {_, :__struct__, 0} -> ["struct"]
+        _ -> []
+      end
+
+    actual_def = actual_def(name, arity, kind)
+
+    %{
+      doc_fallback: fn ->
+        impl = Map.fetch(module_data.impls, actual_def)
+
+        callback_doc_ast(name, arity, impl) ||
+          delegate_doc_ast(metadata[:delegate_to])
+      end,
+      extra_annotations: extra_annotations,
+      line_override: find_function_line(module_data, actual_def),
+      specs: specs(kind, name, actual_def, module_data)
+    }
+  end
+
+  @impl true
+  def callback_data(entry, module_data) do
+    {{kind, name, arity}, _anno, _signature, _doc, _metadata} = entry
+    actual_def = actual_def(name, arity, kind)
+
+    specs =
+      case Map.fetch(module_data.callbacks, actual_def) do
```

Comment:
So it seems the module_data we receive in callbacks is a superset of the one we return. If so we should document the input module_data too. :)

Reply:
Or is this actually `module_node`?

Reply:
oops, it's neither! I added type specs.